keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discrete,
+countable per-node extended resources was added. The kubelet device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources is too limited. This KEP addresses
+limitations of the current approach for the following use cases:
+
+- *Device initialization*: When starting a workload that uses
+  an accelerator like an FPGA, I’d like to have the accelerator
+  reconfigured or reprogrammed for the workload before the workload
+  itself starts. For security reasons, workloads should not be able to
+  reconfigure devices directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+  For the FPGA example, a file containing the desired configuration
+  of the FPGA has to be referenced.
+
+- *Device cleanup*: When my workload is finished, I would like to have
+  a mechanism for cleanup of the device, that will ensure that device
+  does not contain traces/parameters/data from previous workloads and
+  appropriate power state/shutdown. For example, an FPGA might have
+  to be reset because its configuration for the workload was
+  confidential.
+
+  *Limitation*: Post-stop actions are not supported.
+
+- *Partial allocation*: When deploying a container I’d like to be able
+  to use part of the shareable device inside a container and other
+  containers should be able to use other free resources on the same
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow one to allocate part of the device because parameters are too limited
+  and Kubernetes doesn't have enough information about the extended
+  resources on a node to decide whether they can be shared.
+
+- *Optional allocation*: When deploying a workload I’d like to specify
+  soft(optional) device requirements. If a device exists and it’s
+  allocatable it will be allocated. If not - the workload will be run on
+  a node without a device. GPU and crypto-offload engines are
+  examples of this kind of device. If they’re not available, workloads
+  can still run by falling back to using only the CPU for the same
+  task.
+
+  *Limitation*: Optional allocation is supported neither by the device
+  plugins nor by current Pod resource declaration.
+
+- *Support Over the Fabric devices*: When deploying a container, I’d
+  like to utilize devices available over the Fabric (network, special
+  links, etc).
+
+  *Limitation*: Device Plugins framework supports only local devices.
```

Comment:
The existence of a viable workaround is great - it takes some of the pressure off.  That said, I think it's important to think about what we really want and how it is best expressed.

I'm not a big fan of gross hacks, but I am also very conservative about where and how we spend our dwindling budget for complexity.

Reply:
We've discussed the Akri approach further and it's worth pointing out that the "race conditions" that I alluded to above exist: what I meant is that a pod might get scheduled onto a node and then cannot run there because it's resource was taken by some other pod in the meantime. kubelet then will retry running it, but in the meantime any memory or CPU that was set aside for the pod is not available for other pods that otherwise might run on the node.

This KEP avoids such a situation.